### DISCUSSION TITLE
Channels

### DISCUSSION THREAD

**Discussion by Trenly** [Trenly/winget-pkgs#115] [General]
@OfficialEsco @jedieaston 

I plan to look at release channels next.

Heres what I know needs to be implemented -
* Ability to update specific channel only
* Ability to specify channel when creating installers
* Showing channel when using QuickUpdate (Mode 2)
* Ability to specify channel through command line parameters

My main question is - Should there be a default channel? I.E - should we default to “Release” unless it is set otherwise? 

Secondary question - What else do we need to consider for channels?

**Comment by jedieaston** [comment]
First question: my opinion is that there shouldn't be a default channel unless putting the channel in the manifest is required by the schema (it isn't at this stage). If we put "Release" and it turns out 80% of apps use "Stable" in their docs, it'll get confusing.

I don't know if we know enough about the implementation of channels in the client to answer the second question. I don't know if it's even been roadmapped.

**Reply by Trenly** [reply]
Makes sense to me

**Reply by denelon** [reply]
I was thinking about a key to identify the "default" channel for a package so when we run into the case where the version is the same for more than one channel we would know which one is the default channel so we wouldn't install the "latest" version if the default channel isn't there yet. I'm open to other suggestions.

package 1
```
     Version: 1.0
          DefaultChannel: "Release"
          Installers:
               Installer:
                   Channel: "Release"

     Version: 1.1
          DefaultChannel: "Release"
               Installer:
                   Channel: "Preview"
```
package 2
```
     Version 1.0
          DefaultChannel: "Stable"
          Installers:
               Installer:
                    Channel: "Stable"
               Installer:
                     Channel: "Beta"

    Version 1.1
          DefaultChannel: "Stable"
          Installers
               Installer
                    Channel: "Beta"
 ```

**Reply by Trenly** [reply]
I like this approach. Validation of this has a few nuances though, since not every package uses release channels, and as you mentioned before, we have to allow arbitrary values. 

I'm thinking that if any one installer in a manifest has a channel associated with it, then every installer should be required to have a channel and the default channel should also be required

**Reply by denelon** [reply]
Another critical challenge is when the user installed a package outside of the Windows Package Manager and we aren't able to determine the channel. We may need to prompt if it isn't specified, and who knows what might happen if you try to upgrade from one channel to another.

**Reply by Trenly** [reply]
> Another critical challenge is when the user installed a package outside of the Windows Package Manager and we aren't able to determine the channel. We may need to prompt if it isn't specified, and who knows what might happen if you try to upgrade from one channel to another.

I did a bit more searching, and there seems to be discussions about this all over the place. Would it be better for me to start a specification document over in the cli repo just so we can have one central place to talk through and document all of these considerations?

**Reply by denelon** [reply]
That would be very much appreciated! A draft PR would pull everything in one place, and we can update the specification tracker on the issue as well.

**Comment by OfficialEsco** [comment]
Ugh the channel talk, we need @denelon to clarify how to work with the channels..

AFAIK channels are supposed to be for software that ~can~ can't be installed side by side with the same name.
This means Alpha, Dev, Beta and Stable should still be split into different packages.
Stable and Pre-Release ~could MAYBE be 2 different channels? Or does Pre-Release go under the Beta branch?~ where the name is the same can go under the same folder

A prime example for channels are LibreOffice which uses Still (LTSR) and and Fresh (Current) branch.

It would be fantastic if we can use Channels across Packages under the same folder, this would help with packages which keep the same name for Stable and Beta (there is a bunch)
However this is most likely not supported

The second package that comes to mind with a different use case is the Python pacakges

**Reply by jedieaston** [reply]
https://github.com/microsoft/winget-cli/discussions/1445 is the last word about it.

**Reply by Trenly** [reply]
I would disagree about the side by side portion;
See https://github.com/microsoft/winget-cli/discussions/1672#discussioncomment-1665516

Here it seems like they are intended for when it CANT be side by side. The packages which CAN be side by side should get their own ID’s

**Reply by OfficialEsco** [reply]
Yuuuup i remembered it the other way around..
Well, that maybe doesn't change anything that i suggested, i've not tested if the packages with the same Publisher and PackageName can be installed side by side, it makes sense that it overrides itself(?)

After trying to figure out the difference, that might actually be what i accidentally said 🥴

**Comment by vedantmgoyal9** [comment]
This is what I feel would be better:

> My main question is - Should there be a default channel? I.E - should we default to “Release” unless it is set otherwise?

Yes, when nothing is specified, it should be the default channel, i.e. `release`, `stable`, etc...

> Secondary question - What else do we need to consider for channels?

What would the script do when there are channels `release`, `beta`, `nightly` and someone enters "default" or "stable"? I mean to say, would there be any kind of "redirect" or "grouping"?

Example:

Group 1: Stable, Release, Default
Group 2: Beta, Preview, Insider
Group 3: Developer, Nightly

**Reply by vedantmgoyal9** [reply]
@OfficialEsco has also put a similar point (if I am not wrong), he just posted about 10 seconds earlier.

**Reply by Trenly** [reply]
> What would the script do when there are channels release, beta, nightly and someone enters "default" or "stable"? I mean to say, would there be any kind of "redirect" or "grouping"?

This gets a little complex;

New Packages - Specifying “-Channel” switch would apply that channel to all installers

Package Updates - Specifying “-Channel”
- If the channel exists in the last manifest, only update the installers from that Channel
- If the channel does not exist
  - Full Update Mode, notify user channel doesn’t exist, continue as if “-Channel” wasn’t specified
  - Quick Update Mode, Depend on user settings file; Default to prompting for channel which exists in the package; User can have setting to quit
  - Auto Update, notify and exit

**Reply by vedantmgoyal9** [reply]
The approach you suggested can also be used. Except for this, I don't think there is any other thing that needs to be considered as of now.
